python/astro_metadata_translator/translator.py

```diff
@@ -1047,6 +1047,41 @@ def observing_date_to_offset(cls, observing_date: astropy.time.Time) -> astropy.
         """
         return None
 
+    @classmethod
+    def observing_date_to_observing_day(
+        cls, observing_date: astropy.time.Time, offset: astropy.time.TimeDelta | int | None
+    ) -> int:
+        """Return the YYYYMMDD integer corresponding to the observing day.
+
+        The offset is subtracted from the time of observation before
+        calculating the year, month and day.
+
+        Parameters
+        ----------
+        observing_date : `astropy.time.Time`
+            The observation date.
+        offset : `astropy.time.TimeDelta` | int | None
+            The offset to subtract from the observing date when calculating
+            the observing day. If `None` no offset is applied.
+
+        Returns
+        -------
+        day : `int`
+            The observing day as an integer of form YYYYMMDD.
+
+        Notes
+        -----
+        For example, if the offset is +12 hours both 2023-07-06T13:00 and
+        2023-07-07T11:00 will return an observing day of 20230706 because
+        the observing day goes from 2023-07-06T12:00 to 2023-07-07T12:00.
+        """
+        observing_date = observing_date.tai
+        if offset:
+            if isinstance(offset, int):
+                offset = astropy.time.TimeDelta(offset, format="sec", scale="tai")
+            observing_date -= offset
+        return int(observing_date.strftime("%Y%m%d"))
```

Comment:
Is it possible to do it without doing string conversion? This may fail for year 999 if `strftime` adds 0 at the start (but maybe not, and you can add `base=10`).

Reply:
A TAI for year less than 1970-ish is not allowed.

Reply:
It does seem to work fine:
```
>>> t.Time("0999-02-05T12:45:00", format="isot", scale="tai").strftime("%Y%m%d")
'09990205'
>>> int(t.Time("0999-02-05T12:45:00", format="isot", scale="tai").strftime("%Y%m%d"))
9990205
```

I haven't worked out how to get the year/month/day without the string formatting. It only takes 18 microseconds to do the string formatting and cast to int though.

Reply:
My idea was that internally `strftime` already does its job to extract year/month/day, so you can save few microseconds doing just arithmetic instead of string conversion, but if it too hard then strings are OK.